From e0a1c68b44f1da9d7aaec364ab492abde3feca9a Mon Sep 17 00:00:00 2001
From: Nikhil Namjoshi <nikhilnamjoshi@google.com>
Date: Fri, 10 Nov 2023 19:09:53 +0000
Subject: [PATCH] Add generic patching support in rded

Design Doc : go/minibmc-redfish-patch-support
With this change, we should be able to support
patch requests for any miniBMC resource. The implementation
logic is generic to handle encoding of any resource and
properties using libbej encoding
Tested:
Tested by sending live patch requests for different property
types and verified that miniBMC Redfish resource response
shows those types. Verifed the libbej encoding correctness using
unit tests.
Patch Tracking Bug: b/310244088
Upstream info / review: NA
Upstream-Status: Pending
Justification:
The patch depends upon rded design being upstreamed which is still in progress
(https://gerrit.openbmc.org/c/openbmc/docs/+/61256)
Google-Bug-Id: 277137230
Google-Bug-Id: 309736282
Change-Id: I0b7d9d98a096dd3294ee0724ef5f78136fefdf0d
Signed-off-by: Nikhil Namjoshi <nikhilnamjoshi@google.com>
%% original patch: 0010-Add-generic-patching-support-in-rded.patch
---
 rded/helper/rde_operation/rde_operation.cpp   |  57 +-
 .../helper/rde_operation/rde_update_utils.cpp | 341 +++++++++++
 .../helper/rde_operation/rde_update_utils.hpp |  35 ++
 rded/meson.build                              |  11 +
 test/meson.build                              |   5 +-
 test/rde_update_utils_test.cpp                | 579 ++++++++++++++++++
 6 files changed, 971 insertions(+), 57 deletions(-)
 create mode 100644 rded/helper/rde_operation/rde_update_utils.cpp
 create mode 100644 rded/helper/rde_operation/rde_update_utils.hpp
 create mode 100644 test/rde_update_utils_test.cpp

diff --git a/rded/helper/rde_operation/rde_operation.cpp b/rded/helper/rde_operation/rde_operation.cpp
index ab36eeb..03155e2 100644
--- a/rded/helper/rde_operation/rde_operation.cpp
+++ b/rded/helper/rde_operation/rde_operation.cpp
@@ -11,6 +11,7 @@
 #include "json.hpp"
 #include "libbej/bej_decoder_json.hpp"
 #include "libbej/bej_encoder_json.hpp"
+#include "rde_update_utils.hpp"
 
 #include <chrono>
 #include <cstring>
@@ -368,62 +369,6 @@ int encodePayload(const std::string& requestPayload, const std::string& action,
     return 0;
 }
 
-int encodePayloadForUpdateOp(std::string redfishReqPayload,
-                             uint8_t** schemaDict, uint8_t** annotationDict,
-                             std::vector<uint8_t>& encodedPayload)
-{
-    // Nikhil : Currently this encoding is only suitable to encode the patch
-    // operation payload for rde pwm sensors (E.g. Payload {"Reading": 32.0}).
-    // This needs to be made more generic to support patch operation on
-    // different resource and their properites
-    BejDictionaries bejDictionaries;
-    bejDictionaries.schemaDictionary = *schemaDict;
-    bejDictionaries.annotationDictionary = *annotationDict;
-    bejDictionaries.errorDictionary = nullptr;
-
-    // Get rid of spaces from the string
-    redfishReqPayload.erase(
-        std::remove(redfishReqPayload.begin(), redfishReqPayload.end(), ' '),
-        redfishReqPayload.end());
-
-    // Get rid of the outer curly braces from the patch data
-    redfishReqPayload = redfishReqPayload.substr(1,
-                                                 redfishReqPayload.size() - 2);
-
-    // Get rid of '"' from the property name
-    redfishReqPayload.erase(
-        std::remove(redfishReqPayload.begin(), redfishReqPayload.end(), '"'),
-        redfishReqPayload.end());
-
-    size_t delimiterPos = redfishReqPayload.find(":");
-    if (delimiterPos == std::string::npos)
-    {
-        std::cerr << "Delimiter ':' absent in the Redfish Req Data"
-                  << std::endl;
-        return -1;
-    }
-
-    std::string propertyName = redfishReqPayload.substr(0, delimiterPos);
-    double propertyValue =
-        std::stof(redfishReqPayload.substr(delimiterPos + 1));
-
-    struct RedfishPropertyParent parent;
-    bejTreeInitSet(&parent, nullptr);
-
-    struct RedfishPropertyLeafReal child;
-    bejTreeAddReal(&parent, &child, propertyName.c_str(), propertyValue);
-
-    libbej::BejEncoderJson encoder;
-    int rc = encoder.encode(&bejDictionaries, bejMajorSchemaClass, &parent);
-    if (rc)
-    {
-        std::cerr << "Encoding failure with rc: " << rc << std::endl;
-        return rc;
-    }
-    encodedPayload = encoder.getOutput();
-    return 0;
-}
-
 int processRdeOperationInit(int fd, uint8_t eid, int instanceId,
                             struct pldm_rde_requester_manager* manager,
                             struct pldm_rde_requester_context* ctx,
diff --git a/rded/helper/rde_operation/rde_update_utils.cpp b/rded/helper/rde_operation/rde_update_utils.cpp
new file mode 100644
index 0000000..69b0627
--- /dev/null
+++ b/rded/helper/rde_operation/rde_update_utils.cpp
@@ -0,0 +1,341 @@
+/**
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "libbej/bej_common.h"
+#include "libbej/bej_dictionary.h"
+
+#include "libbej/bej_encoder_json.hpp"
+#include "nlohmann/json.hpp"
+#include "rde_update_utils.hpp"
+
+#include <cstring>
+#include <iostream>
+#include <sstream>
+#include <unordered_map>
+#include <variant>
+
+using json = nlohmann::json;
+using BejChildNodesVariant =
+    std::variant<struct RedfishPropertyParent, struct RedfishPropertyLeafNull,
+                 struct RedfishPropertyLeafInt, struct RedfishPropertyLeafEnum,
+                 struct RedfishPropertyLeafString,
+                 struct RedfishPropertyLeafReal,
+                 struct RedfishPropertyLeafBool>;
+
+static int addLeafToBejTree(
+    const struct BejDictionaryProperty* property,
+    const std::string& propertyName, json& propertyValue,
+    struct RedfishPropertyParent* root,
+    std::unordered_map<std::string, BejChildNodesVariant>& childNodesStorage)
+{
+    switch (property->format.principalDataType)
+    {
+        case bejNull:
+        {
+            struct RedfishPropertyLeafNull child;
+            childNodesStorage[propertyName] = child;
+            bejTreeAddNull(root,
+                           &std::get<struct RedfishPropertyLeafNull>(
+                               childNodesStorage[propertyName]),
+                           propertyName.c_str());
+            break;
+        }
+        case bejInteger:
+        {
+            try
+            {
+                int64_t data = propertyValue.get<int64_t>();
+                std::cerr << "Converted integer value: " << data << std::endl;
+                struct RedfishPropertyLeafInt child;
+                childNodesStorage[propertyName] = child;
+                bejTreeAddInteger(root,
+                                  &std::get<struct RedfishPropertyLeafInt>(
+                                      childNodesStorage[propertyName]),
+                                  propertyName.c_str(), data);
+            }
+            catch (const json::exception& e)
+            {
+                std::cerr << "Integer Conversion error: " << e.what()
+                          << std::endl;
+                return -1;
+            }
+            break;
+        }
+        case bejEnum:
+        {
+            try
+            {
+                // Libbej APIs accept char* pointer for strings. So we need to
+                // grab the json string value by reference as a local string
+                // variable can get deallocated.
+                std::string& data = propertyValue.get_ref<std::string&>();
+                std::cerr << "Converted enum value: " << data << std::endl;
+                struct RedfishPropertyLeafEnum child;
+                childNodesStorage[propertyName] = child;
+                bejTreeAddEnum(root,
+                               &std::get<struct RedfishPropertyLeafEnum>(
+                                   childNodesStorage[propertyName]),
+                               propertyName.c_str(), data.c_str());
+            }
+            catch (const json::exception& e)
+            {
+                std::cerr << "Enum Conversion error: " << e.what() << std::endl;
+                return -1;
+            }
+            break;
+        }
+        case bejString:
+        {
+            try
+            {
+                // Libbej APIs accept char* pointer for strings. So we need to
+                // grab the json string value by reference as a local string
+                // variable can get deallocated
+                std::string& data = propertyValue.get_ref<std::string&>();
+                std::cerr << "Converted string value: " << data << std::endl;
+                struct RedfishPropertyLeafString child;
+                childNodesStorage[propertyName] = child;
+                bejTreeAddString(root,
+                                 &std::get<struct RedfishPropertyLeafString>(
+                                     childNodesStorage[propertyName]),
+                                 propertyName.c_str(), data.c_str());
+            }
+            catch (const json::exception& e)
+            {
+                std::cerr << "String Conversion error: " << e.what()
+                          << std::endl;
+                return -1;
+            }
+            break;
+        }
+        case bejReal:
+        {
+            try
+            {
+                double data = propertyValue.get<double>();
+                std::cerr << "Converted real/double value: " << data
+                          << std::endl;
+                struct RedfishPropertyLeafReal child;
+                childNodesStorage[propertyName] = child;
+                bejTreeAddReal(root,
+                               &std::get<struct RedfishPropertyLeafReal>(
+                                   childNodesStorage[propertyName]),
+                               propertyName.c_str(), data);
+            }
+            catch (const json::exception& e)
+            {
+                std::cerr << "Real/Double Conversion error: " << e.what()
+                          << std::endl;
+                return -1;
+            }
+            break;
+        }
+        case bejBoolean:
+        {
+            try
+            {
+                bool data = propertyValue.get<bool>();
+                std::cerr << "Converted boolean value: " << data << std::endl;
+                struct RedfishPropertyLeafBool child;
+                childNodesStorage[propertyName] = child;
+                bejTreeAddBool(root,
+                               &std::get<struct RedfishPropertyLeafBool>(
+                                   childNodesStorage[propertyName]),
+                               propertyName.c_str(), data);
+            }
+            catch (const json::exception& e)
+            {
+                std::cerr << "Boolean Conversion error: " << e.what()
+                          << std::endl;
+                return -1;
+            }
+            break;
+        }
+        default:
+            std::cerr << "Unknown Bej Data type " << std::endl;
+            return -1;
+    }
+    return 0;
+}
+
+static int addBranchToBejTree(
+    const uint8_t* schemaDict, uint16_t startingPropertyOffset,
+    const std::string& propertyName, json& propertyValue,
+    struct RedfishPropertyParent* root,
+    std::unordered_map<std::string, BejChildNodesVariant>& childNodesStorage)
+{
+    uint16_t currentOffset;
+    const struct BejDictionaryProperty* property = nullptr;
+
+    // Get the property details from the schema dictionary to identify the
+    // property type
+    int rc = bejDictGetPropertyByName(schemaDict, startingPropertyOffset,
+                                      propertyName.c_str(), &property,
+                                      &currentOffset);
+    if (rc != 0)
+    {
+        std::cerr << "Failed to find dictionary sequence for property "
+                  << propertyName << " for the resource, rc : " << rc
+                  << std::endl;
+        return rc;
+    }
+    uint16_t childPointerOffset = property->childPointerOffset;
+
+    if (property->format.principalDataType == bejPropertyAnnotation)
+    {
+        // As per Redfish specifiction, patching is not allowed on the Odata
+        // annotation properties. So return error in this case.
+        std::cerr
+            << "It is an annotation property. Patching is not allowed on annotated properties"
+            << std::endl;
+        return -1;
+    }
+    else if (property->format.principalDataType == bejSet)
+    {
+        // If it is a set then link it with the parent and continue encoding the
+        // rest of the BEJ tree
+
+        struct RedfishPropertyParent child;
+        childNodesStorage[propertyName] = child;
+        bejTreeInitSet(&std::get<struct RedfishPropertyParent>(
+                           childNodesStorage[propertyName]),
+                       propertyName.c_str());
+
+        // Since it is a set, link the child to the parent and continue
+        // recursively fetching the leaf property
+        bejTreeLinkChildToParent(root, &std::get<struct RedfishPropertyParent>(
+                                           childNodesStorage[propertyName]));
+
+        // Iterate through the JSON object.
+        // There could be multiple properties that are being patched at a time
+        // and they can be present on different branches of the BEJ tree. So
+        // construct the BEJ tree branches at each level
+        for (const auto& it : propertyValue.items())
+        {
+            const std::string& childMemberPropertyName = it.key();
+            std::cerr << "Property name: " << childMemberPropertyName
+                      << std::endl;
+
+            json& childMemberPropertyValue = it.value();
+
+            rc = addBranchToBejTree(schemaDict, childPointerOffset,
+                                    childMemberPropertyName,
+                                    childMemberPropertyValue,
+                                    &std::get<struct RedfishPropertyParent>(
+                                        childNodesStorage[propertyName]),
+                                    childNodesStorage);
+            if (rc != 0)
+            {
+                return rc;
+            }
+        }
+    }
+    else
+    {
+        // If it is a leaf property then this is a property that has to be
+        // updated/patched. The recursion ends when we find a leaf property
+        rc = addLeafToBejTree(property, propertyName, propertyValue, root,
+                              childNodesStorage);
+        if (rc != 0)
+        {
+            return rc;
+        }
+    }
+
+    return 0;
+}
+
+int encodePayloadForUpdateOp(const std::string& redfishReqPayload,
+                             uint8_t** schemaDict, uint8_t** annotationDict,
+                             std::vector<uint8_t>& encodedPayload)
+{
+    if (redfishReqPayload.empty())
+    {
+        return -1;
+    }
+
+    if ((schemaDict == nullptr) || (*schemaDict == nullptr))
+    {
+        std::cerr << "Schema dictionary cannot be NULL" << std::endl;
+        return -1;
+    }
+
+    if ((annotationDict == nullptr) || (*annotationDict == nullptr))
+    {
+        std::cerr << "Annotation dictionary cannot be NULL" << std::endl;
+        return -1;
+    }
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = *schemaDict;
+    bejDictionaries.annotationDictionary = *annotationDict;
+    bejDictionaries.errorDictionary = nullptr;
+    int rc;
+
+    struct RedfishPropertyParent parent;
+    bejTreeInitSet(&parent, nullptr);
+    json redfishReqJsonPayload;
+
+    // Get the Patch Data in JSON format
+    try
+    {
+        redfishReqJsonPayload = json::parse(redfishReqPayload);
+    }
+    catch (const json::parse_error& e)
+    {
+        std::cerr << "JSON parsing error: " << e.what()
+                  << " while parsing sensor collection response "
+                  << redfishReqPayload << std::endl;
+        return -1;
+    }
+
+    uint16_t rootOffset = bejDictGetPropertyHeadOffset();
+
+    // Create a storage for the bej child nodes so that their memory is not
+    // deallocated by the time we use the libbej encoder to encode the tree in
+    // to a binary.
+    std::unordered_map<std::string, BejChildNodesVariant> childNodesStorage;
+
+    // Iterate through the JSON object
+    // There could be multiple properties that are being patched at a time
+    // and they can be present on different branches of the BEJ tree. So
+    // construct the BEJ tree branches at each level
+    for (const auto& it : redfishReqJsonPayload.items())
+    {
+        const std::string& memberPropertyName = it.key();
+        std::cerr << "Property name: " << memberPropertyName << std::endl;
+        json& memberPropertyValue = it.value();
+
+        // Create a BEJ tree for encoding
+        rc = addBranchToBejTree(bejDictionaries.schemaDictionary, rootOffset,
+                                memberPropertyName, memberPropertyValue,
+                                &parent, childNodesStorage);
+        if (rc != 0)
+        {
+            return rc;
+        }
+    }
+
+    libbej::BejEncoderJson encoder;
+    rc = encoder.encode(&bejDictionaries, bejMajorSchemaClass, &parent);
+    if (rc != 0)
+    {
+        std::cerr << "Encoding failure with rc: " << rc << std::endl;
+        return rc;
+    }
+    encodedPayload = encoder.getOutput();
+    return 0;
+}
diff --git a/rded/helper/rde_operation/rde_update_utils.hpp b/rded/helper/rde_operation/rde_update_utils.hpp
new file mode 100644
index 0000000..a2131fa
--- /dev/null
+++ b/rded/helper/rde_operation/rde_update_utils.hpp
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+/**
+ * @brief Encodes the Patch request payload using libbej encoding
+ *
+ * @param[in] redfishReqPayload JSON Request Payload
+ * @param[in] schemaDict uint8_t double pointer to schema dictionary array
+ * @param[in] annotationDict uint8_t double pointer to annotation dictionary
+ * array
+ * @param[inout] encodedPayload encodedPayload vector that will store the libbej
+ * encoded request payload
+ *
+ * @return bool (True for error response and false otherwise)
+ */
+int encodePayloadForUpdateOp(const std::string& redfishReqPayload,
+                             uint8_t** schemaDict, uint8_t** annotationDict,
+                             std::vector<uint8_t>& encodedPayload);
diff --git a/rded/meson.build b/rded/meson.build
index 16c90d8..4de00b8 100644
--- a/rded/meson.build
+++ b/rded/meson.build
@@ -3,6 +3,7 @@ sources = [
   'helper/discovery/base_discovery.cpp',
   'helper/discovery/rde_discovery.cpp',
   'helper/rde_operation/rde_operation.cpp',
+  'helper/rde_operation/rde_update_utils.cpp'
 ]
 
 args = [
@@ -46,6 +47,16 @@ deps = [
   stdplus,
   req_src,
 ]
+
+librdedupdatelib_headers = ['.']
+librdedupdatelib = static_library(
+  'rdedupdatelib',
+  'helper/rde_operation/rde_update_utils.cpp',
+  include_directories : include_directories(librdedupdatelib_headers),
+  implicit_include_directories: false,
+  dependencies: deps
+)
+
 executable(
   'rded',
   sources,
diff --git a/test/meson.build b/test/meson.build
index b88162c..e850ce3 100644
--- a/test/meson.build
+++ b/test/meson.build
@@ -4,7 +4,8 @@ test_src = declare_dependency(
 
 tests = [
   'pldmd_registration_test',
-  'rde_utils_test'
+  'rde_utils_test',
+  'rde_update_utils_test'
 ]
 
 foreach t : tests
@@ -12,8 +13,10 @@ foreach t : tests
                      implicit_include_directories: false,
                      link_args: dynamic_linker,
                      build_rpath: get_option('oe-sdk').allowed() ? rpath : '',
+                     link_with: librdedupdatelib,
                      dependencies: [
                          libpldm_dep,
+                         libbej_dep,
                          nlohmann_json_dep,
                          sdbusplus,
                          boost,
diff --git a/test/rde_update_utils_test.cpp b/test/rde_update_utils_test.cpp
new file mode 100644
index 0000000..099f443
--- /dev/null
+++ b/test/rde_update_utils_test.cpp
@@ -0,0 +1,579 @@
+#include "libbej/bej_decoder_json.hpp"
+#include "libbej/bej_encoder_json.hpp"
+#include "nlohmann/json.hpp"
+#include "rded/helper/rde_operation/rde_update_utils.hpp"
+
+#include <array>
+#include <cstdint>
+#include <span>
+#include <string>
+#include <string_view>
+
+#include <gmock/gmock-matchers.h>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+using ::testing::_;
+using ::testing::Eq;
+using ::testing::MatcherCast;
+using ::testing::Pointee;
+using ::testing::Return;
+using ::testing::SafeMatcherCast;
+using ::testing::StrEq;
+
+std::vector<uint8_t> redfish_annotation_dictionary = {
+    0x0,  0x0,  0x4a, 0x0,  0x0,  0xf0, 0xf0, 0xf1, 0xf0, 0x6,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x16, 0x0,  0x13, 0x0,  0xc,  0xf0, 0x2,  0x10, 0x0,
+    0x0,  0xd4, 0x0,  0x1,  0x0,  0x16, 0xfc, 0x2,  0x50, 0x1,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x14, 0x12, 0x3,  0x10, 0x2,  0x0,  0x2e, 0x1,  0x1,
+    0x0,  0x19, 0x26, 0x3,  0x0,  0x3,  0x0,  0x38, 0x1,  0x2,  0x0,  0x20,
+    0x3f, 0x3,  0x0,  0x4,  0x0,  0x4c, 0x1,  0x2,  0x0,  0x1b, 0x5f, 0x3,
+    0x40, 0x5,  0x0,  0x60, 0x1,  0x5,  0x0,  0x1c, 0x7a, 0x3,  0x0,  0x6,
+    0x0,  0x92, 0x1,  0x4,  0x0,  0x23, 0x96, 0x3,  0x70, 0x7,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x1a, 0xb9, 0x3,  0x70, 0x8,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x1a, 0xd3, 0x3,  0x70, 0x9,  0x0,  0x0,  0x0,  0x0,  0x0,  0x19,
+    0xed, 0x3,  0x0,  0xa,  0x0,  0xba, 0x1,  0x6,  0x0,  0x12, 0x6,  0x4,
+    0x0,  0xb,  0x0,  0xf6, 0x1,  0x3,  0x0,  0x1b, 0x18, 0x4,  0x70, 0xc,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x1e, 0x33, 0x4,  0x50, 0xd,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0xf,  0x51, 0x4,  0x30, 0xe,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0xd,  0x60, 0x4,  0x50, 0xf,  0x0,  0x0,  0x0,  0x0,  0x0,  0xc,
+    0x6d, 0x4,  0x50, 0x10, 0x0,  0x0,  0x0,  0x0,  0x0,  0xa,  0x79, 0x4,
+    0x50, 0x11, 0x0,  0x0,  0x0,  0x0,  0x0,  0x10, 0x83, 0x4,  0x50, 0x12,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0xc,  0x93, 0x4,  0x0,  0x0,  0x0,  0xde,
+    0x0,  0x8,  0x0,  0x0,  0x0,  0x0,  0x52, 0x0,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x8,  0x9f, 0x4,  0x12, 0x1,  0x0,  0x2e, 0x1,  0x1,  0x0,  0xc,
+    0xa7, 0x4,  0x52, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0xa,  0xb3, 0x4,
+    0x0,  0x3,  0x0,  0x0,  0x0,  0x0,  0x0,  0x4,  0xbd, 0x4,  0x12, 0x4,
+    0x0,  0x2e, 0x1,  0x1,  0x0,  0x12, 0xc1, 0x4,  0x52, 0x5,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0xb,  0xd3, 0x4,  0x52, 0x6,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x9,  0xde, 0x4,  0x42, 0x7,  0x0,  0x14, 0x2,  0x3,  0x0,  0x10,
+    0xe7, 0x4,  0x50, 0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,
+    0x10, 0x0,  0x0,  0x32, 0x2,  0x1,  0x0,  0xd,  0xf7, 0x4,  0x32, 0x1,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0xb,  0x4,  0x5,  0x30, 0x0,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x23, 0xf,  0x5,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x1b, 0x32, 0x5,  0x50, 0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x19,
+    0x4d, 0x5,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0xa,  0x66, 0x5,
+    0x50, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0x1b, 0x70, 0x5,  0x50, 0x3,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x8,  0x8b, 0x5,  0x50, 0x4,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x15, 0x93, 0x5,  0x32, 0x0,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x23, 0xf,  0x5,  0x2,  0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0x1a,
+    0xa8, 0x5,  0x52, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0x1b, 0x32, 0x5,
+    0x12, 0x3,  0x0,  0x5a, 0x2,  0x1,  0x0,  0x10, 0xc2, 0x5,  0x56, 0x0,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x5,  0xd2, 0x5,  0x10, 0x1,  0x0,  0xd4,
+    0x0,  0x1,  0x0,  0x9,  0xd7, 0x5,  0x2,  0x2,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0xf,  0xe0, 0x5,  0x56, 0x3,  0x0,  0x0,  0x0,  0x0,  0x0,  0x5,
+    0xef, 0x5,  0x12, 0x4,  0x0,  0x64, 0x2,  0x1,  0x0,  0x14, 0xf4, 0x5,
+    0x2,  0x5,  0x0,  0x0,  0x0,  0x0,  0x0,  0x1a, 0xa8, 0x5,  0x40, 0x0,
+    0x0,  0x6e, 0x2,  0x4,  0x0,  0xa,  0x8,  0x6,  0x30, 0x1,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x23, 0xf,  0x5,  0x50, 0x2,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x1b, 0x32, 0x5,  0x50, 0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x9,
+    0x12, 0x6,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0x3,  0x1b, 0x6,
+    0x50, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0x8,  0x1e, 0x6,  0x0,  0x0,
+    0x0,  0x3c, 0x2,  0x3,  0x0,  0x0,  0x0,  0x0,  0x2,  0x0,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x13, 0x26, 0x6,  0x0,  0x1,  0x0,  0x96, 0x2,  0x3,
+    0x0,  0x6,  0x39, 0x6,  0x42, 0x2,  0x0,  0xb4, 0x2,  0x5,  0x0,  0x8,
+    0x3f, 0x6,  0x40, 0x0,  0x0,  0x60, 0x1,  0x5,  0x0,  0x0,  0x0,  0x0,
+    0x40, 0x0,  0x0,  0x6e, 0x2,  0x4,  0x0,  0x0,  0x0,  0x0,  0x50, 0x0,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x19, 0x4d, 0x5,  0x50, 0x1,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0xa,  0x66, 0x5,  0x50, 0x2,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x1b, 0x70, 0x5,  0x50, 0x3,  0x0,  0x0,  0x0,  0x0,  0x0,  0x8,
+    0x8b, 0x5,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x4,  0xbd, 0x4,
+    0x16, 0x1,  0x0,  0xe6, 0x2,  0x1,  0x0,  0xc,  0x47, 0x6,  0x2,  0x2,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x11, 0x53, 0x6,  0x50, 0x0,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x1a, 0x64, 0x6,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0xf,  0x7e, 0x6,  0x50, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0x25,
+    0x8d, 0x6,  0x50, 0x3,  0x0,  0x0,  0x0,  0x0,  0x0,  0x19, 0xb2, 0x6,
+    0x50, 0x4,  0x0,  0x0,  0x0,  0x0,  0x0,  0x24, 0xcb, 0x6,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x41, 0x6e, 0x6e, 0x6f,
+    0x74, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x73, 0x0,  0x40, 0x4d, 0x65, 0x73,
+    0x73, 0x61, 0x67, 0x65, 0x2e, 0x45, 0x78, 0x74, 0x65, 0x6e, 0x64, 0x65,
+    0x64, 0x49, 0x6e, 0x66, 0x6f, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69,
+    0x73, 0x68, 0x2e, 0x41, 0x63, 0x74, 0x69, 0x6f, 0x6e, 0x49, 0x6e, 0x66,
+    0x6f, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69, 0x73, 0x68, 0x2e, 0x41,
+    0x6c, 0x6c, 0x6f, 0x77, 0x61, 0x62, 0x6c, 0x65, 0x56, 0x61, 0x6c, 0x75,
+    0x65, 0x73, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69, 0x73, 0x68, 0x2e,
+    0x43, 0x6f, 0x6c, 0x6c, 0x65, 0x63, 0x74, 0x69, 0x6f, 0x6e, 0x43, 0x61,
+    0x70, 0x61, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x69, 0x65, 0x73, 0x0,  0x40,
+    0x52, 0x65, 0x64, 0x66, 0x69, 0x73, 0x68, 0x2e, 0x4d, 0x61, 0x69, 0x6e,
+    0x74, 0x65, 0x6e, 0x61, 0x6e, 0x63, 0x65, 0x57, 0x69, 0x6e, 0x64, 0x6f,
+    0x77, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69, 0x73, 0x68, 0x2e, 0x4f,
+    0x70, 0x65, 0x72, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x41, 0x70, 0x70, 0x6c,
+    0x79, 0x54, 0x69, 0x6d, 0x65, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69,
+    0x73, 0x68, 0x2e, 0x4f, 0x70, 0x65, 0x72, 0x61, 0x74, 0x69, 0x6f, 0x6e,
+    0x41, 0x70, 0x70, 0x6c, 0x79, 0x54, 0x69, 0x6d, 0x65, 0x53, 0x75, 0x70,
+    0x70, 0x6f, 0x72, 0x74, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69, 0x73,
+    0x68, 0x2e, 0x4f, 0x70, 0x74, 0x69, 0x6f, 0x6e, 0x61, 0x6c, 0x4f, 0x6e,
+    0x43, 0x72, 0x65, 0x61, 0x74, 0x65, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66,
+    0x69, 0x73, 0x68, 0x2e, 0x52, 0x65, 0x71, 0x75, 0x69, 0x72, 0x65, 0x64,
+    0x4f, 0x6e, 0x43, 0x72, 0x65, 0x61, 0x74, 0x65, 0x0,  0x40, 0x52, 0x65,
+    0x64, 0x66, 0x69, 0x73, 0x68, 0x2e, 0x53, 0x65, 0x74, 0x4f, 0x6e, 0x6c,
+    0x79, 0x4f, 0x6e, 0x43, 0x72, 0x65, 0x61, 0x74, 0x65, 0x0,  0x40, 0x52,
+    0x65, 0x64, 0x66, 0x69, 0x73, 0x68, 0x2e, 0x53, 0x65, 0x74, 0x74, 0x69,
+    0x6e, 0x67, 0x73, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66, 0x69, 0x73, 0x68,
+    0x2e, 0x53, 0x65, 0x74, 0x74, 0x69, 0x6e, 0x67, 0x73, 0x41, 0x70, 0x70,
+    0x6c, 0x79, 0x54, 0x69, 0x6d, 0x65, 0x0,  0x40, 0x52, 0x65, 0x64, 0x66,
+    0x69, 0x73, 0x68, 0x2e, 0x55, 0x70, 0x64, 0x61, 0x74, 0x61, 0x62, 0x6c,
+    0x65, 0x41, 0x66, 0x74, 0x65, 0x72, 0x43, 0x72, 0x65, 0x61, 0x74, 0x65,
+    0x0,  0x40, 0x6f, 0x64, 0x61, 0x74, 0x61, 0x2e, 0x63, 0x6f, 0x6e, 0x74,
+    0x65, 0x78, 0x74, 0x0,  0x40, 0x6f, 0x64, 0x61, 0x74, 0x61, 0x2e, 0x63,
+    0x6f, 0x75, 0x6e, 0x74, 0x0,  0x40, 0x6f, 0x64, 0x61, 0x74, 0x61, 0x2e,
+    0x65, 0x74, 0x61, 0x67, 0x0,  0x40, 0x6f, 0x64, 0x61, 0x74, 0x61, 0x2e,
+    0x69, 0x64, 0x0,  0x40, 0x6f, 0x64, 0x61, 0x74, 0x61, 0x2e, 0x6e, 0x65,
+    0x78, 0x74, 0x4c, 0x69, 0x6e, 0x6b, 0x0,  0x40, 0x6f, 0x64, 0x61, 0x74,
+    0x61, 0x2e, 0x74, 0x79, 0x70, 0x65, 0x0,  0x4d, 0x65, 0x73, 0x73, 0x61,
+    0x67, 0x65, 0x0,  0x4d, 0x65, 0x73, 0x73, 0x61, 0x67, 0x65, 0x41, 0x72,
+    0x67, 0x73, 0x0,  0x4d, 0x65, 0x73, 0x73, 0x61, 0x67, 0x65, 0x49, 0x64,
+    0x0,  0x4f, 0x65, 0x6d, 0x0,  0x52, 0x65, 0x6c, 0x61, 0x74, 0x65, 0x64,
+    0x50, 0x72, 0x6f, 0x70, 0x65, 0x72, 0x74, 0x69, 0x65, 0x73, 0x0,  0x52,
+    0x65, 0x73, 0x6f, 0x6c, 0x75, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x53, 0x65,
+    0x76, 0x65, 0x72, 0x69, 0x74, 0x79, 0x0,  0x4d, 0x65, 0x73, 0x73, 0x61,
+    0x67, 0x65, 0x53, 0x65, 0x76, 0x65, 0x72, 0x69, 0x74, 0x79, 0x0,  0x43,
+    0x61, 0x70, 0x61, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x69, 0x65, 0x73, 0x0,
+    0x4d, 0x61, 0x78, 0x4d, 0x65, 0x6d, 0x62, 0x65, 0x72, 0x73, 0x0,  0x4d,
+    0x61, 0x69, 0x6e, 0x74, 0x65, 0x6e, 0x61, 0x6e, 0x63, 0x65, 0x57, 0x69,
+    0x6e, 0x64, 0x6f, 0x77, 0x44, 0x75, 0x72, 0x61, 0x74, 0x69, 0x6f, 0x6e,
+    0x49, 0x6e, 0x53, 0x65, 0x63, 0x6f, 0x6e, 0x64, 0x73, 0x0,  0x4d, 0x61,
+    0x69, 0x6e, 0x74, 0x65, 0x6e, 0x61, 0x6e, 0x63, 0x65, 0x57, 0x69, 0x6e,
+    0x64, 0x6f, 0x77, 0x53, 0x74, 0x61, 0x72, 0x74, 0x54, 0x69, 0x6d, 0x65,
+    0x0,  0x41, 0x74, 0x4d, 0x61, 0x69, 0x6e, 0x74, 0x65, 0x6e, 0x61, 0x6e,
+    0x63, 0x65, 0x57, 0x69, 0x6e, 0x64, 0x6f, 0x77, 0x53, 0x74, 0x61, 0x72,
+    0x74, 0x0,  0x49, 0x6d, 0x6d, 0x65, 0x64, 0x69, 0x61, 0x74, 0x65, 0x0,
+    0x49, 0x6e, 0x4d, 0x61, 0x69, 0x6e, 0x74, 0x65, 0x6e, 0x61, 0x6e, 0x63,
+    0x65, 0x57, 0x69, 0x6e, 0x64, 0x6f, 0x77, 0x4f, 0x6e, 0x52, 0x65, 0x73,
+    0x65, 0x74, 0x0,  0x4f, 0x6e, 0x52, 0x65, 0x73, 0x65, 0x74, 0x0,  0x4f,
+    0x6e, 0x53, 0x74, 0x61, 0x72, 0x74, 0x55, 0x70, 0x64, 0x61, 0x74, 0x65,
+    0x52, 0x65, 0x71, 0x75, 0x65, 0x73, 0x74, 0x0,  0x4d, 0x61, 0x69, 0x6e,
+    0x74, 0x65, 0x6e, 0x61, 0x6e, 0x63, 0x65, 0x57, 0x69, 0x6e, 0x64, 0x6f,
+    0x77, 0x52, 0x65, 0x73, 0x6f, 0x75, 0x72, 0x63, 0x65, 0x0,  0x53, 0x75,
+    0x70, 0x70, 0x6f, 0x72, 0x74, 0x65, 0x64, 0x56, 0x61, 0x6c, 0x75, 0x65,
+    0x73, 0x0,  0x45, 0x54, 0x61, 0x67, 0x0,  0x4d, 0x65, 0x73, 0x73, 0x61,
+    0x67, 0x65, 0x73, 0x0,  0x53, 0x65, 0x74, 0x74, 0x69, 0x6e, 0x67, 0x73,
+    0x4f, 0x62, 0x6a, 0x65, 0x63, 0x74, 0x0,  0x54, 0x69, 0x6d, 0x65, 0x0,
+    0x53, 0x75, 0x70, 0x70, 0x6f, 0x72, 0x74, 0x65, 0x64, 0x41, 0x70, 0x70,
+    0x6c, 0x79, 0x54, 0x69, 0x6d, 0x65, 0x73, 0x0,  0x41, 0x70, 0x70, 0x6c,
+    0x79, 0x54, 0x69, 0x6d, 0x65, 0x0,  0x43, 0x72, 0x69, 0x74, 0x69, 0x63,
+    0x61, 0x6c, 0x0,  0x4f, 0x4b, 0x0,  0x57, 0x61, 0x72, 0x6e, 0x69, 0x6e,
+    0x67, 0x0,  0x43, 0x61, 0x70, 0x61, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x69,
+    0x65, 0x73, 0x4f, 0x62, 0x6a, 0x65, 0x63, 0x74, 0x0,  0x4c, 0x69, 0x6e,
+    0x6b, 0x73, 0x0,  0x55, 0x73, 0x65, 0x43, 0x61, 0x73, 0x65, 0x0,  0x52,
+    0x65, 0x6c, 0x61, 0x74, 0x65, 0x64, 0x49, 0x74, 0x65, 0x6d, 0x0,  0x54,
+    0x61, 0x72, 0x67, 0x65, 0x74, 0x43, 0x6f, 0x6c, 0x6c, 0x65, 0x63, 0x74,
+    0x69, 0x6f, 0x6e, 0x0,  0x43, 0x6f, 0x6d, 0x70, 0x75, 0x74, 0x65, 0x72,
+    0x53, 0x79, 0x73, 0x74, 0x65, 0x6d, 0x43, 0x6f, 0x6d, 0x70, 0x6f, 0x73,
+    0x69, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x56, 0x6f, 0x6c, 0x75, 0x6d, 0x65,
+    0x43, 0x72, 0x65, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x43, 0x6f, 0x6d,
+    0x70, 0x75, 0x74, 0x65, 0x72, 0x53, 0x79, 0x73, 0x74, 0x65, 0x6d, 0x43,
+    0x6f, 0x6e, 0x73, 0x74, 0x72, 0x61, 0x69, 0x6e, 0x65, 0x64, 0x43, 0x6f,
+    0x6d, 0x70, 0x6f, 0x73, 0x69, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x52, 0x65,
+    0x73, 0x6f, 0x75, 0x72, 0x63, 0x65, 0x42, 0x6c, 0x6f, 0x63, 0x6b, 0x43,
+    0x6f, 0x6d, 0x70, 0x6f, 0x73, 0x69, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x52,
+    0x65, 0x73, 0x6f, 0x75, 0x72, 0x63, 0x65, 0x42, 0x6c, 0x6f, 0x63, 0x6b,
+    0x43, 0x6f, 0x6e, 0x73, 0x74, 0x72, 0x61, 0x69, 0x6e, 0x65, 0x64, 0x43,
+    0x6f, 0x6d, 0x70, 0x6f, 0x73, 0x69, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x0};
+
+std::vector<uint8_t> redfish_control_dictionary = {
+    0x0,  0x1,  0x1b, 0x0,  0x0,  0xf0, 0xf0, 0xf1, 0x31, 0x2,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x16, 0x0,  0xa,  0x0,  0x8,  0x1a, 0x1,  0x4,  0x7,
+    0x0,  0x7a, 0x0,  0x4,  0x0,  0xc,  0x22, 0x1,  0x44, 0x8,  0x0,  0xa2,
+    0x0,  0x4,  0x0,  0xc,  0x2e, 0x1,  0x46, 0x9,  0x0,  0xca, 0x0,  0x3,
+    0x0,  0xc,  0x3a, 0x1,  0x56, 0xc,  0x0,  0x0,  0x0,  0x0,  0x0,  0xc,
+    0x46, 0x1,  0x52, 0xd,  0x0,  0x0,  0x0,  0x0,  0x0,  0x3,  0x52, 0x1,
+    0x46, 0xe,  0x0,  0xe8, 0x0,  0x3,  0x0,  0xf,  0x55, 0x1,  0x52, 0x11,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x5,  0x64, 0x1,  0x64, 0x19, 0x0,  0x0,
+    0x0,  0x0,  0x0,  0x9,  0x69, 0x1,  0x46, 0x1a, 0x0,  0x6,  0x1,  0x2,
+    0x0,  0xd,  0x72, 0x1,  0x56, 0x1b, 0x0,  0x0,  0x0,  0x0,  0x0,  0xe,
+    0x7f, 0x1,  0x56, 0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x16, 0x8d, 0x1,
+    0x64, 0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0xd,  0xa3, 0x1,  0x64, 0x2,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x9,  0xb0, 0x1,  0x64, 0x3,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0xd,  0xb9, 0x1,  0x50, 0x0,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0xa,  0xc6, 0x1,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0x9,
+    0xd0, 0x1,  0x50, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0x7,  0xd9, 0x1,
+    0x50, 0x3,  0x0,  0x0,  0x0,  0x0,  0x0,  0x9,  0xe0, 0x1,  0x50, 0x0,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0xa,  0xe9, 0x1,  0x50, 0x1,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x6,  0xf3, 0x1,  0x50, 0x2,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0xc,  0xf9, 0x1,  0x50, 0x0,  0x0,  0x0,  0x0,  0x0,  0x0,  0x7,
+    0x5,  0x2,  0x50, 0x1,  0x0,  0x0,  0x0,  0x0,  0x0,  0xa,  0xc,  0x2,
+    0x50, 0x2,  0x0,  0x0,  0x0,  0x0,  0x0,  0xd,  0x16, 0x2,  0x50, 0x0,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x6,  0x23, 0x2,  0x50, 0x1,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x7,  0x29, 0x2,  0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f,
+    0x6c, 0x0,  0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x4c, 0x6f, 0x6f,
+    0x70, 0x0,  0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x4d, 0x6f, 0x64,
+    0x65, 0x0,  0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x54, 0x79, 0x70,
+    0x65, 0x0,  0x44, 0x65, 0x73, 0x63, 0x72, 0x69, 0x70, 0x74, 0x69, 0x6f,
+    0x6e, 0x0,  0x49, 0x64, 0x0,  0x49, 0x6d, 0x70, 0x6c, 0x65, 0x6d, 0x65,
+    0x6e, 0x74, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x0,  0x4e, 0x61, 0x6d, 0x65,
+    0x0,  0x53, 0x65, 0x74, 0x50, 0x6f, 0x69, 0x6e, 0x74, 0x0,  0x53, 0x65,
+    0x74, 0x50, 0x6f, 0x69, 0x6e, 0x74, 0x54, 0x79, 0x70, 0x65, 0x0,  0x53,
+    0x65, 0x74, 0x50, 0x6f, 0x69, 0x6e, 0x74, 0x55, 0x6e, 0x69, 0x74, 0x73,
+    0x0,  0x43, 0x6f, 0x65, 0x66, 0x66, 0x69, 0x63, 0x69, 0x65, 0x6e, 0x74,
+    0x55, 0x70, 0x64, 0x61, 0x74, 0x65, 0x54, 0x69, 0x6d, 0x65, 0x0,  0x44,
+    0x69, 0x66, 0x66, 0x65, 0x72, 0x65, 0x6e, 0x74, 0x69, 0x61, 0x6c, 0x0,
+    0x49, 0x6e, 0x74, 0x65, 0x67, 0x72, 0x61, 0x6c, 0x0,  0x50, 0x72, 0x6f,
+    0x70, 0x6f, 0x72, 0x74, 0x69, 0x6f, 0x6e, 0x61, 0x6c, 0x0,  0x41, 0x75,
+    0x74, 0x6f, 0x6d, 0x61, 0x74, 0x69, 0x63, 0x0,  0x44, 0x69, 0x73, 0x61,
+    0x62, 0x6c, 0x65, 0x64, 0x0,  0x4d, 0x61, 0x6e, 0x75, 0x61, 0x6c, 0x0,
+    0x4f, 0x76, 0x65, 0x72, 0x72, 0x69, 0x64, 0x65, 0x0,  0x46, 0x72, 0x65,
+    0x71, 0x75, 0x65, 0x6e, 0x63, 0x79, 0x0,  0x50, 0x6f, 0x77, 0x65, 0x72,
+    0x0,  0x54, 0x65, 0x6d, 0x70, 0x65, 0x72, 0x61, 0x74, 0x75, 0x72, 0x65,
+    0x0,  0x44, 0x69, 0x72, 0x65, 0x63, 0x74, 0x0,  0x4d, 0x6f, 0x6e, 0x69,
+    0x74, 0x6f, 0x72, 0x65, 0x64, 0x0,  0x50, 0x72, 0x6f, 0x67, 0x72, 0x61,
+    0x6d, 0x6d, 0x61, 0x62, 0x6c, 0x65, 0x0,  0x52, 0x61, 0x6e, 0x67, 0x65,
+    0x0,  0x53, 0x69, 0x6e, 0x67, 0x6c, 0x65, 0x0,  0x0};
+
+std::vector<uint8_t> redfish_update_service_dictionary = {
+    0x0,  0x1,  0x6,  0x0,  0x0,  0xf0, 0x10, 0xf1, 0x92, 0x0,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x16, 0x0,  0x5,  0x0,  0xe,  0x48, 0x0,  0x2,  0x2,
+    0x0,  0x0,  0x0,  0x0,  0x0,  0x12, 0x56, 0x0,  0x52, 0x3,  0x0,  0x0,
+    0x0,  0x0,  0x0,  0x3,  0x68, 0x0,  0x52, 0x4,  0x0,  0x0,  0x0,  0x0,
+    0x0,  0x5,  0x6b, 0x0,  0x74, 0x6,  0x0,  0x0,  0x0,  0x0,  0x0,  0xf,
+    0x70, 0x0,  0x36, 0xe,  0x0,  0x0,  0x0,  0x0,  0x0,  0x12, 0x7f, 0x0,
+    0x55, 0x70, 0x64, 0x61, 0x74, 0x65, 0x53, 0x65, 0x72, 0x76, 0x69, 0x63,
+    0x65, 0x0,  0x46, 0x69, 0x72, 0x6d, 0x77, 0x61, 0x72, 0x65, 0x49, 0x6e,
+    0x76, 0x65, 0x6e, 0x74, 0x6f, 0x72, 0x79, 0x0,  0x49, 0x64, 0x0,  0x4e,
+    0x61, 0x6d, 0x65, 0x0,  0x53, 0x65, 0x72, 0x76, 0x69, 0x63, 0x65, 0x45,
+    0x6e, 0x61, 0x62, 0x6c, 0x65, 0x64, 0x0,  0x4d, 0x61, 0x78, 0x49, 0x6d,
+    0x61, 0x67, 0x65, 0x53, 0x69, 0x7a, 0x65, 0x42, 0x79, 0x74, 0x65, 0x73,
+    0x0,  0x0};
+
+class TestRdeUpdateUtils : public ::testing::Test
+{
+  public:
+    TestRdeUpdateUtils() {}
+
+    ~TestRdeUpdateUtils() {}
+};
+
+TEST_F(TestRdeUpdateUtils, inputValidationTest)
+{
+    const std::string redfishReqPayload = "";
+    std::vector<uint8_t> encodedPayload;
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    // Pass an empty request payload and verify that the API returns error
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqPayload, &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                -1);
+    const std::string redfishReqControlPayload = "\"Id\": \"HDD_Setpoint\"";
+
+    // Pass a NULL argument for schema dictionary
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlPayload, NULL,
+                                         &annotation_dict_ptr, encodedPayload),
+                -1);
+
+    uint8_t* schemaDict = NULL;
+    // Pass a pointer to NULL pointer of uint8_t type for schemaDict
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlPayload, &schemaDict,
+                                         &annotation_dict_ptr, encodedPayload),
+                -1);
+
+    // Pass a NULL argument for annotation dictionary
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlPayload,
+                                         &control_dict_ptr, NULL,
+                                         encodedPayload),
+                -1);
+
+    uint8_t* annotationDict = NULL;
+    // Pass a pointer to NULL pointer of uint8_t type for annotationDict
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlPayload,
+                                         &control_dict_ptr, &annotationDict,
+                                         encodedPayload),
+                -1);
+}
+
+// Try encoding a root level property of type bejString
+TEST_F(TestRdeUpdateUtils, encodeBejStringDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = control_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqControlStringPayload = "{\"Id\":\"HDD_Setpoint\"}";
+    nlohmann::json redfishReqControlStringPayloadJson =
+        nlohmann::json::parse(redfishReqControlStringPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlStringPayload,
+                                         &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("Id") != decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("Id").value(), "HDD_Setpoint");
+}
+
+// Try encoding a property of type bejReal one level above root
+TEST_F(TestRdeUpdateUtils, encodeBejRealDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = control_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqControlRealPayload =
+        "{\"ControlLoop\" : {\"Differential\" : 10.2}}";
+    nlohmann::json redfishReqControlRealPayloadJson =
+        nlohmann::json::parse(redfishReqControlRealPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlRealPayload,
+                                         &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("ControlLoop") !=
+                decodedPayloadJson.end());
+
+    nlohmann::json controlLoopJson =
+        decodedPayloadJson.find("ControlLoop").value();
+
+    EXPECT_EQ(controlLoopJson.size(), 1);
+
+    EXPECT_EQ(controlLoopJson.find("Differential").value(), 10.2);
+}
+
+// Try encoding a property of type bejEnum one level above root
+TEST_F(TestRdeUpdateUtils, encodeBejEnumDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = control_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqControlEnumPayload =
+        "{\"ControlMode\" : \"Automatic\"}";
+    nlohmann::json redfishReqControlEnumPayloadJson =
+        nlohmann::json::parse(redfishReqControlEnumPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlEnumPayload,
+                                         &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("ControlMode") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("ControlMode").value(), "Automatic");
+}
+
+// Try encoding a property of type bejString, bejRel and bejEnum on multiple
+// levels in json
+TEST_F(TestRdeUpdateUtils, encodeBejEnumStringRealDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = control_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqControlPayload =
+        "{\"Id\":\"HDD_Setpoint\", \"ControlMode\" : \"Automatic\", \"ControlLoop\" : {\"Differential\" : 10.2}}";
+    nlohmann::json redfishReqControlPayloadJson =
+        nlohmann::json::parse(redfishReqControlPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlPayload,
+                                         &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 3);
+
+    EXPECT_TRUE(decodedPayloadJson.find("Id") != decodedPayloadJson.end());
+    EXPECT_EQ(decodedPayloadJson.find("Id").value(), "HDD_Setpoint");
+
+    EXPECT_TRUE(decodedPayloadJson.find("ControlMode") !=
+                decodedPayloadJson.end());
+    EXPECT_EQ(decodedPayloadJson.find("ControlMode").value(), "Automatic");
+
+    EXPECT_TRUE(decodedPayloadJson.find("ControlLoop") !=
+                decodedPayloadJson.end());
+    nlohmann::json controlLoopJson =
+        decodedPayloadJson.find("ControlLoop").value();
+    EXPECT_EQ(controlLoopJson.size(), 1);
+    EXPECT_EQ(controlLoopJson.find("Differential").value(), 10.2);
+}
+
+// Try encoding a root level property of type bejInt
+TEST_F(TestRdeUpdateUtils, encodeBejIntDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* update_service_dict_ptr = redfish_update_service_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = update_service_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqUpdateServiceIntPayload =
+        "{\"MaxImageSizeBytes\": 4536754}";
+    nlohmann::json redfishReqUpdateServiceIntPayloadJson =
+        nlohmann::json::parse(redfishReqUpdateServiceIntPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqUpdateServiceIntPayload,
+                                         &update_service_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("MaxImageSizeBytes") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("MaxImageSizeBytes").value(), 4536754);
+}
+
+// Try encoding a root level property of type bejBool
+TEST_F(TestRdeUpdateUtils, encodeBejBoolDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* update_service_dict_ptr = redfish_update_service_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = update_service_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqUpdateServiceBoolPayload =
+        "{\"ServiceEnabled\": true}";
+    nlohmann::json redfishReqUpdateServiceBoolPayloadJson =
+        nlohmann::json::parse(redfishReqUpdateServiceBoolPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqUpdateServiceBoolPayload,
+                                         &update_service_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("ServiceEnabled") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("ServiceEnabled").value(), true);
+}
+
+// Try encoding a root level property of type bejInt and bejBool
+TEST_F(TestRdeUpdateUtils, encodeBejIntBoolDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* update_service_dict_ptr = redfish_update_service_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = update_service_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqUpdateServicePayload =
+        "{\"MaxImageSizeBytes\": 4536754, \"ServiceEnabled\": true}";
+    nlohmann::json redfishReqUpdateServicePayloadJson =
+        nlohmann::json::parse(redfishReqUpdateServicePayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqUpdateServicePayload,
+                                         &update_service_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 2);
+
+    EXPECT_TRUE(decodedPayloadJson.find("MaxImageSizeBytes") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("MaxImageSizeBytes").value(), 4536754);
+
+    EXPECT_TRUE(decodedPayloadJson.find("ServiceEnabled") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("ServiceEnabled").value(), true);
+}
+
+// Try encoding a property of type bejNull
+TEST_F(TestRdeUpdateUtils, encodeBejNullDataTypeTest)
+{
+    uint8_t* annotation_dict_ptr = redfish_annotation_dictionary.data();
+    uint8_t* control_dict_ptr = redfish_control_dictionary.data();
+
+    BejDictionaries bejDictionaries;
+    bejDictionaries.schemaDictionary = control_dict_ptr;
+    bejDictionaries.annotationDictionary = annotation_dict_ptr;
+    bejDictionaries.errorDictionary = nullptr;
+
+    std::string redfishReqControlNullPayload = "{\"ControlMode\" : null}";
+    nlohmann::json redfishReqControlNullPayloadJson =
+        nlohmann::json::parse(redfishReqControlNullPayload);
+    std::vector<uint8_t> encodedPayload;
+    EXPECT_THAT(encodePayloadForUpdateOp(redfishReqControlNullPayload,
+                                         &control_dict_ptr,
+                                         &annotation_dict_ptr, encodedPayload),
+                0);
+
+    libbej::BejDecoderJson decoder;
+    decoder.decode(bejDictionaries, encodedPayload);
+    std::string decodedString = decoder.getOutput();
+    std::cerr << "decodedString : " << decodedString << std::endl;
+    nlohmann::json decodedPayloadJson = nlohmann::json::parse(decodedString);
+
+    EXPECT_EQ(decodedPayloadJson.size(), 1);
+
+    EXPECT_TRUE(decodedPayloadJson.find("ControlMode") !=
+                decodedPayloadJson.end());
+
+    EXPECT_EQ(decodedPayloadJson.find("ControlMode").value(), nullptr);
+}
-- 
2.34.1

